### PR TITLE
#15844 adding language support and refactor

### PR DESCRIPTION
**Pull request by jdotcms**

### CODE REVIEW

**Review by jgambarios — APPROVED** [review]

File: dotCMS/src/main/java/com/dotmarketing/servlets/ShortyServlet.java [hunk 1]
```diff
@@ -1,188 +1,284 @@
...
-  private void _serve(final HttpServletRequest request, final HttpServletResponse response) throws Exception {
 
-    Host host = WebAPILocator.getHostWebAPI().getCurrentHost(request);
+  private int getWeight (final String uri, final int defaultWeight) {
 
+    int weight = 0;
 
-    PageMode mode = PageMode.get(request);
+    try {
+      final Matcher weightMatcher = weightPattern.matcher(uri);
```
Comment (wezell) [inline dotCMS/src/main/java/com/dotmarketing/servlets/ShortyServlet.java]: It is not weight, it is width

File: dotCMS/src/main/java/com/dotmarketing/servlets/ShortyServlet.java [hunk 2]
```diff
@@ -1,188 +1,284 @@
...
-    try{
-      Matcher m = dimw.matcher(uri);
-      w = m.find() ? Integer.parseInt(m.group().substring(1).replace("w", "")) : w;
-      m = dimh.matcher(uri);
-      h = m.find() ? Integer.parseInt(m.group().substring(1).replace("h", "")) : h;
-    }
-    catch(Exception e){
-      Logger.debug(this, e.getMessage());
-      // let this one die''
+    final String identifier           = tokens.nextToken();
```
Comment (wezell) [inline dotCMS/src/main/java/com/dotmarketing/servlets/ShortyServlet.java]: The shorty takes an inode or an identifier which are very different

Comment (wezell) [reply dotCMS/src/main/java/com/dotmarketing/servlets/ShortyServlet.java]: We only need to do multilingual work if what we pass in is an identifier

File: dotCMS/src/main/java/com/dotmarketing/servlets/ShortyServlet.java [hunk 3]
```diff
@@ -1,188 +1,284 @@
...
+                                   final String tryField,
+                                   final boolean live)
             throws DotStateException, DotDataException {
 
-        final Optional<Field> fieldOpt = resolveField(con, tryField);
+        final Optional<Field> fieldOpt = resolveField(contentlet, tryField);
 
         if (!fieldOpt.isPresent()) {
-            return "/" + con.getInode() + "/" + FILE_ASSEST_DEFAULT;
+            return new StringBuilder(StringPool.FORWARD_SLASH).append(contentlet.getInode())
```
Comment (wezell) [inline dotCMS/src/main/java/com/dotmarketing/servlets/ShortyServlet.java]: This is unnecessary optimization - the compiler takes care of string concatenation and we lose readability

File: dotCMS/src/main/java/com/dotmarketing/servlets/ShortyServlet.java [hunk 4]
```diff
@@ -1,188 +1,284 @@
...
+    final ShortyId shorty  = shortOpt.get();
+    final String   path    = isImage? "/contentAsset/image" : "/contentAsset/raw-data";
+    final User systemUser  = APILocator.systemUser();
+    try {
 
-    if (!shortOpt.isPresent()) {
+      final Contentlet contentlet = shorty.type == ShortType.IDENTIFIER ?
+              this.contentletAPI.findContentletByIdentifier
+                      (shorty.longId, live,
+                              this.getLanguageId(request.getParameter(WebKeys.LANGUAGE_ID_PARAMETER)),
```
Comment (wezell) [inline dotCMS/src/main/java/com/dotmarketing/servlets/ShortyServlet.java]: We need to handle default language and language fallback here.

**Review by wezell — COMMENTED** [review]
See comments

File: dotCMS/src/main/java/com/dotmarketing/servlets/ShortyServlet.java (hunk 4 above)
Comment (jdotcms) [reply dotCMS/src/main/java/com/dotmarketing/servlets/ShortyServlet.java]: Done

File: dotCMS/src/main/java/com/dotmarketing/servlets/ShortyServlet.java [hunk 5]
```diff
@@ -1,188 +1,284 @@
...
+
+    if (!live) {
+      response.setHeader("Pragma", "no-cache");
+      response.setHeader("Cache-Control", "no-cache");
+      response.setDateHeader("Expires", 0);
     }
   }
 
+  private boolean isValidRequest(final HttpServletRequest request,
+                                 final HttpServletResponse response, PageMode mode) throws DotDataException, DotSecurityException, PortalException, SystemException, IOException {
```
[the remark this answers is not in the document]
Comment (jdotcms) [reply dotCMS/src/main/java/com/dotmarketing/servlets/ShortyServlet.java]: done

File: dotCMS/src/main/java/com/dotmarketing/servlets/ShortyServlet.java (hunk 3 above)
Comment (jdotcms) [reply dotCMS/src/main/java/com/dotmarketing/servlets/ShortyServlet.java]: done

**Pull request merged**